### PR TITLE
[receiver/sqlquery] fix handling of nulls

### PR DESCRIPTION
**Pull request by pmcollins**
**Description:** This fixes the receiver panicking on NULL values. Additional refactoring was done to improve test coverage.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19177

**Testing:** Unit tests, integration tests, and manual tests passed.

**Documentation:** Docs added about NULL handling.

### CODE REVIEW

File: receiver/sqlqueryreceiver/README.md [hunk 1]
```diff
@@ -96,12 +101,20 @@ Data point attributes:
 Value: 1
 ```
 
+#### NULL values
+
+Avoid queries that produce any NULL values. If a query produces a NULL value, a warning will be logged. Furthermore,
+if a configuration references the column that produces a NULL value, an additional error will be logged. However, in
+either case, the recevier will continue to operate.
```
Comment (pmcollins) [inline receiver/sqlqueryreceiver/README.md:108]: I intend on fixing this typo after initial reviews (don't want to trigger a recheck on everything yet).

File: receiver/sqlqueryreceiver/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+func TestSetDataPointValue(t *testing.T) {
+	err := setDataPointValue(MetricCfg{
+		ValueType:   MetricValueTypeInt,
+		ValueColumn: "some-col",
+	}, "", pmetric.NewNumberDataPoint())
+	assert.Error(t, err)
```
Comment (rmfitzpatrick) [inline receiver/sqlqueryreceiver/metrics_test.go:29]: Would it make sense to add an error content check or is it potentially in flux?

File: receiver/sqlqueryreceiver/row_scanner.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+func newRowScanner(colTypes []colType) *rowScanner {
+	rs := &rowScanner{
+		cols: map[string]func() (string, error){},
+	}
+	for _, sqlType := range colTypes {
+		colName := sqlType.Name()
+		var v any
+		rs.cols[colName] = func() (string, error) {
+			if v == nil {
```
Comment (rmfitzpatrick) [inline receiver/sqlqueryreceiver/row_scanner.go:40]: What would the implications be for using the special `"<nil>"` string instead?

File: receiver/sqlqueryreceiver/db_client_test.go [hunk 4]
```diff
@@ -14,19 +14,159 @@
...
+		"col_2": "goodbye",
+		"col_3": "false",
+		"col_4": "4.2",
+	}, rows[1])
+}
+
+func TestDBSQLClient_Nulls(t *testing.T) {
+	cl := dbSQLClient{
+		db: fakeDB{rowVals: [][]any{
+			{42, nil}, // NULLs from the DB map to nil here
```
Comment (rmfitzpatrick) [inline receiver/sqlqueryreceiver/db_client_test.go:78]: Do you mind adding additional values to show row processing continues?

File: receiver/sqlqueryreceiver/row_scanner.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+				return "", errNullValueWarning
+			}
+			format := "%v"
+			if reflect.TypeOf(v).Kind() == reflect.Slice {
+				// The Postgres driver returns a []uint8 (a string) for decimal and numeric types,
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply receiver/sqlqueryreceiver/row_scanner.go]: Fixed.

File: receiver/sqlqueryreceiver/metrics_test.go (hunk 2 above)
Comment (pmcollins) [reply receiver/sqlqueryreceiver/metrics_test.go]: Good idea, fixed.

**Review by rmfitzpatrick — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Pull request merged**